File: app/src/main/java/com/mohsenoid/rickandmorty/ui/base/BaseFragment.java
package com.mohsenoid.rickandmorty.ui.base;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.mohsenoid.rickandmorty.RickAndMortyApplication;
import com.mohsenoid.rickandmorty.injection.DependenciesProvider;

public abstract class BaseFragment extends Fragment {

    @Override
    public void onAttach(@NonNull Context context) {
        DependenciesProvider dependenciesProvider = ((RickAndMortyApplication) context.getApplicationContext()).getDependenciesProvider();

        injectDependencies(dependenciesProvider);
        super.onAttach(context);
    }

    public abstract void injectDependencies(DependenciesProvider dependenciesProvider);
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/view/base/BaseFragment.kt
package com.mohsenoid.rickandmorty.view.base

import android.content.Context
import androidx.fragment.app.Fragment
import com.mohsenoid.rickandmorty.util.dispatcher.DispatcherProvider
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import org.koin.android.ext.android.inject
import kotlin.coroutines.CoroutineContext

abstract class BaseFragment : Fragment(), CoroutineScope {

    private val dispatcherProvider: DispatcherProvider by inject()

    private lateinit var job: Job
    override lateinit var coroutineContext: CoroutineContext

    override fun onAttach(context: Context) {
        job = SupervisorJob()
        coroutineContext = dispatcherProvider.mainDispatcher + job

        super.onAttach(context)
    }
}
